docs: correct spec docs of fswap module (#1419)
Finschia/finschia-sdk: x/fswap/spec/01_concepts.md

<!--
order: 1
-->

# Concepts

> ## Swap

Reviewer: Providing detailed information would be appreciated:

1. Who can trigger a swap, by which message.
2. Abstract behavior about swap (burn or lock, etc).
3. descriptions about the fields in the messages or references to protobuf so one easily access the description.

The `x/fswap` module defines a `Swap` type in which a coin is allowed to be swapped into another coin on the chain.

```go
type Swap struct {
  FromDenom             string
  ToDenom               string
  AmountCapForToDenom   sdk.Int
  SwapRate              sdk.Dec
}
```

## Config

The `x/fswap` module defines a `Config` type for managing the maximum number of Swaps allowed on chain  through `MaxSwaps`. Additionally, `UpdateAllowed` specifies whether `Swap` can be modified.

```go
type Config struct {
	MaxSwaps      int
	UpdateAllowed bool
}
```

## Proposal

> Typically, a `Swap` is proposed and submitted through foundation via a `MsgSetSwap`.

Reviewer: `x/foundation` or foundation module. Also, it's just an example of `authority`.
`MsgSetSwap` is NOT a proposal, it's an ordinary message like the others (e.g. x/bank MsgSend). Other modules (x/gov, x/foundation) will include the message in their proposals.

Author: > it's just an example of authority.

Could you tell more detail about this? Is it meaning that `MsgSetSwap` could be sent without `authority` or `x/foundation`?
I think it only could be sent by x/foundation not x/gov.

Reviewer: As of now, x/gov cannot control x/swap. By design, however, it could be the authority. You may think following enhancements:

1. x/gov message based proposal update
2. removal of x/foundation restrictions on its message based proposals

So, it would be good idea to provide the general concepts first, and giving the current detailed state (it is "concepts" section).

This proposal prescribes to the standard foundation process. If the proposal passes, the `Swap` can be used on chain.

```go
type MsgSetSwap struct {
	Authority           string
    Swap                Swap

>     ToDenomMetadata     bank.Metadata

Reviewer: It would be nice to describe about the restrictions on setting denom metadata (or reference to them).

<!--
order: 1
-->

# Concepts

## Swap

The `x/fswap` module defines a `Swap` type in which a coin is allowed to be swapped into another coin on the chain.
You could find detailed information in the [Protobuf reference](../../../proto/lbm/fswap/v1/fswap.proto#L9-L16) 

```go
type Swap struct {
  FromDenom             string
  ToDenom               string
  AmountCapForToDenom   sdk.Int
  SwapRate              sdk.Dec
}
```

Anyone could use one of the following two transcations to swap `FromDedenom` to `ToDenom`.
1. `simd tx fswap swap [from] [from_coin_amount] [to_denom]`
    - this transcation could swap a specified amount of `from_denom` via [`MsgSwap`](../../../proto/lbm/fswap/v1/tx.proto#L17-L24)
2. `simd tx fswap swap-all [from_address] [from_denom] [to_denom]`
    - this transcation could swap all of `from_denom` under `from_address` via [`MsgSwapAll`](../../../proto/lbm/fswap/v1/tx.proto#L28-L33)

When the swap is triggered, the following event will occur:
1. `from_denom` will be sent from `from_address` to `x/fswap` module
2. `x/fswap` module will burn `from_denom`

> 3. `x/fswap` module will mint `to_denom` of equivalent value

Reviewer 2: How about specifying multiplication like below.
```suggestion
3. `x/fswap` module will mint `to_denom` as amount as `from_denom * swapRate`
```

4. these `to_denom` will sent to `from_address`
5. `EventSwapCoins` will be emitted

## Config

The `x/fswap` module defines a `Config` type for managing the maximum number of Swaps allowed on chain  through `MaxSwaps`. Additionally, `UpdateAllowed` specifies whether `Swap` can be modified.

```go
type Config struct {
	MaxSwaps      int
	UpdateAllowed bool
}
```

## MsgSetSwap

> `MsgSetSwap` is not a proposal; it is a regular message. Other modules can include `MsgSetSwap` in their proposals to set `Swap`. If the proposal passes, the `Swap` can be used on chain.

Reviewer 2: It is a little bit confusing, It says `MsgSetSwap is not proposal` then `If the proposal passes`.
How about this `MsgSetSwap is a message-based proposal`?

Author: [As @0Tech said ](https://github.com/Finschia/finschia-sdk/pull/1419#discussion_r1637493076), MsgSetSwap is not a proposal. 
I think `the proposal passes` here refers to `Other modules can include MsgSetSwap in their proposals` mentioned above.

  AmountCapForToDenom   sdk.Int
  SwapRate              sdk.Dec
}
```

Anyone could use one of the following two transcations to swap `FromDedenom` to `ToDenom`.
1. `simd tx fswap swap [from] [from_coin_amount] [to_denom]`
    - this transcation could swap a specified amount of `from_denom` via [`MsgSwap`](../../../proto/lbm/fswap/v1/tx.proto#L17-L24)
2. `simd tx fswap swap-all [from_address] [from_denom] [to_denom]`
    - this transcation could swap all of `from_denom` under `from_address` via [`MsgSwapAll`](../../../proto/lbm/fswap/v1/tx.proto#L28-L33)

When the swap is triggered, the following event will occur:
1. `from_denom` will be sent from `from_address` to `x/fswap` module
2. `x/fswap` module will burn `from_denom`
3. `x/fswap` module will mint `to_denom` as amount as `from_denom * swapRate`
4. these `to_denom` will sent to `from_address`
5. `EventSwapCoins` will be emitted

## Config

The `x/fswap` module defines a `Config` type for managing the maximum number of Swaps allowed on chain through `MaxSwaps`. Additionally, `UpdateAllowed` specifies whether `Swap` can be modified.

```go
type Config struct {
	MaxSwaps      int
	UpdateAllowed bool
}
```

## MsgSetSwap

> `MsgSetSwap` is not a proposal; it is a regular message. Other modules can include `MsgSetSwap` in their proposals to set `Swap`. If the proposal passes, the `Swap` can be used on chain.

Reviewer: Explicitly stating `MsgSetSwap` is not a proposal sounds awkward, but not critical.

}
```

## MsgSetSwap

Typically, a `Swap` is proposed and submitted through `x/foundation` via a `MsgSetSwap`.

> Other modules (`x/gov`, `x/foundation`) will include this message in their proposals. If the proposal passes, the `Swap` can be used on chain.

Reviewer: We should not mention the exact module name `x/gov` now, because it's not being supported yet.

> `ToDenomMetadata` is [`Metadata`](../../bank/types/bank.pb.go#L325) in `x/bank` module, and it MUST meet these [limitations](../../bank/types/metadata.go#L11).

Reviewer: +1 for the reference to the x/bank denom metadata restrictions. You should also state the restrictions by x/swap.

Reviewer: I mean, we should state the restrictions by x/swap, i.e. one cannot override existing denom metadata, etc.

Author: > one cannot override existing denom metadata

This can be modified by config, doesn't it?

Reviewer: No, it's not allowed. You should also state about the configuration.
